sql/catalog/lease: TestLeaseManagerLockedTimestampConcurrent failed

sql/catalog/lease.TestLeaseManagerLockedTimestampConcurrent [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21169718?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21169718?buildTab=artifacts#/) on release-26.1 @ [83aae1f1935a5beea888bd3a0780437ed9d21b3a](https://github.com/cockroachdb/cockroach/commits/83aae1f1935a5beea888bd3a0780437ed9d21b3a):


```
=== RUN   TestLeaseManagerLockedTimestampConcurrent
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/5632b52c1c67b61307e1d8a1ca2dd18d/logTestLeaseManagerLockedTimestampConcurrent3399340271
    test_log_scope.go:82: use -show-logs to present logs inline
--- FAIL: TestLeaseManagerLockedTimestampConcurrent (140.67s)
panic: pq: context canceled
	(1) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:783
	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease_test.TestLeaseManagerLockedTimestampConcurrent.func1.2
	  | 	pkg/sql/catalog/lease/lease_test.go:3883
	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease_test.TestLeaseManagerLockedTimestampConcurrent.func1.Group.GoCtx.5
	  | 	pkg/util/ctxgroup/ctxgroup.go:200
	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
	  | runtime.goexit
	  | 	src/runtime/asm_arm64.s:1268
	Wraps: (2) pq: context canceled
	Error types: (1) *withstack.withStack (2) *pq.Error [recovered, repanicked]

goroutine 87955 [running]:
testing.tRunner.func1.2({0x4d63e20, 0x400783e8b0})
	GOROOT/src/testing/testing.go:1872 +0x190
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1875 +0x31c
panic({0x4d63e20?, 0x400783e8b0?})
	GOROOT/src/runtime/panic.go:783 +0x120
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x400d991618?, {0x605c750, 0x9d40520})
	pkg/util/stop/stopper.go:226 +0x68
panic({0x4d63e20?, 0x400783e8b0?})
	GOROOT/src/runtime/panic.go:783 +0x120
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0x400a470870, {0x605c750, 0x9d40520})
	pkg/util/stop/stopper.go:568 +0x14c
panic({0x4d63e20?, 0x400783e8b0?})
	GOROOT/src/runtime/panic.go:783 +0x120
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait({0x4007c11b80?, {0x605c9d0?, 0x40010a8c80?}, 0x4007c01d58?})
	pkg/util/ctxgroup/ctxgroup.go:151 +0xc0
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease_test.TestLeaseManagerLockedTimestampConcurrent.func1(0x400791cc40, 0x0)
	pkg/sql/catalog/lease/lease_test.go:3910 +0x45c
github.com/cockroachdb/cockroach/pkg/testutils.RunValues[...].func1()
	pkg/testutils/subtest.go:23 +0x28
testing.tRunner(0x400791cc40, 0x400b459320)
	GOROOT/src/testing/testing.go:1934 +0xc8
created by testing.(*T).Run in goroutine 87934
	GOROOT/src/testing/testing.go:1997 +0x364
=== RUN   TestLeaseManagerLockedTimestampConcurrent/oldVersionRetention=false
    test_server_shim.go:668: DRPC is enabled (via override by TestingGlobalDRPCOption)
--- FAIL: TestLeaseManagerLockedTimestampConcurrent/oldVersionRetention=false (140.67s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseManagerLockedTimestampConcurrent.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-60861

--- a/pkg/sql/catalog/lease/lease_test.go
+++ b/pkg/sql/catalog/lease/lease_test.go
@@ -3830,7 +3830,11 @@ func TestLeaseManagerLockedTimestampConcurrent(t *testing.T) {
 		nextObjectToRead := make(chan string)
 		nextObjectToModify := make(chan string)
 		var objectModified atomic.Bool
-
+		// Detect cancellation from either the client or server.
+		isCancellationError := func(err error) bool {
+			return errors.Is(err, context.Canceled) ||
+				sqltestutils.IsClientSideQueryCanceledErr(err)
+		}
 		// Creates tables in the background.
 		createThreads := func(ctx context.Context) (err error) {
 			defer close(nextObjectToRead)
@@ -3839,7 +3843,7 @@ func TestLeaseManagerLockedTimestampConcurrent(t *testing.T) {
 				objectName := fmt.Sprintf("t%d", i)
 				sql := fmt.Sprintf("CREATE TABLE %s(n int PRIMARY KEY)\n", objectName)
 				if _, err := conn.ExecContext(ctx, sql); err != nil {
-					if errors.Is(err, context.Canceled) {
+					if isCancellationError(err) {
 						return nil
 					}
 					panic(err)
@@ -3876,7 +3880,7 @@ func TestLeaseManagerLockedTimestampConcurrent(t *testing.T) {
 					// Repeat usage of the descriptor.
 					sql := fmt.Sprintf("SELECT * FROM %s", objectName)
 					if _, err := conn.ExecContext(ctx, sql); err != nil {
-						if errors.Is(err, context.Canceled) {
+						if isCancellationError(err) {
 							return nil
 						}
 						panic(err)
@@ -3893,7 +3897,7 @@ func TestLeaseManagerLockedTimestampConcurrent(t *testing.T) {
 			for objectName := range nextObjectToModify {
 				sql := fmt.Sprintf("ALTER TABLE %s ADD COLUMN n2 int", objectName)
 				if _, err := conn.ExecContext(ctx, sql); err != nil {
-					if errors.Is(err, context.Canceled) {
+					if isCancellationError(err) {
 						return nil
 					}
 					panic(err)

